forc-pkg/src/source/reg/mod.rs

```diff
@@ -0,0 +1,523 @@
+            }
+            fs::create_dir_all(&path)?;
+
+            let package_entry = index_file.get(&pinned.source.version).ok_or_else(|| {
+                anyhow!(
+                    "Version {} not found for {}",
+                    pinned.source.version,
+                    pinned.source.name
+                )
+            })?;
+
+            let cid = Cid::from_str(&package_entry.source_cid).with_context(|| {
+                format!(
+                    "Invalid CID {}v{}: `{}`",
+                    package_entry.name, package_entry.version, package_entry.source_cid
+                )
+            })?;
+            match ipfs_node {
+                IPFSNode::Local => {
+                    println_action_green("Fetching", "with local IPFS node");
+                    cid.fetch_with_client(&ipfs_client(), &path).await?;
+                }
+                IPFSNode::WithUrl(gateway_url) => {
+                    println_action_green(
+                        "Fetching",
+                        &format!("from {}. Note: This can take several minutes.", gateway_url),
+                    );
+                    cid.fetch_with_gateway_url(gateway_url, &path).await?;
+                }
+            }
+
+            Ok(path)
+        },
+    )
+    .await?;
+    Ok(path)
+}
+
+async fn with_tmp_fetch_index<F, O, Fut>(
+    fetch_id: u64,
+    pkg_name: &str,
+    source: &Source,
+    f: F,
+) -> anyhow::Result<O>
+where
+    F: FnOnce(IndexFile) -> Fut,
+    Fut: std::future::Future<Output = anyhow::Result<O>>,
+{
+    let tmp_dir = tmp_registry_package_dir(fetch_id, pkg_name, &source.version, &source.namespace);
+    if tmp_dir.exists() {
+        let _ = std::fs::remove_dir_all(&tmp_dir);
+    }
+
+    // TODO: remove the clone
+    let github_resolver = GithubRegistryResolver::with_default_github(source.namespace.clone());
+
+    let path = format!(
+        "{}",
+        location_from_root(github_resolver.chunk_size, &source.namespace, pkg_name).display()
+    );
+    let index_repo_owner = github_resolver.repo_org();
+    let index_repo_name = github_resolver.repo_name();
+    let reference = format!("refs/heads/{}", github_resolver.branch_name());
+    let github_endpoint = format!(
+        "https://raw.githubusercontent.com/{index_repo_owner}/{index_repo_name}/{reference}/{path}"
+    );
+    let client = reqwest::Client::new();
+    let index_response = client.get(github_endpoint).send().await.map_err(|e| {
+        anyhow!("Failed to send request to github to obtain package index file from registry {e}")
+    })?;
+
+    let contents = index_response.text().await?;
+    let index_file: IndexFile = serde_json::from_str(&contents).with_context(|| {
+        format!(
+            "Unable to deserialize a github registry lookup response. Body was: \"{}\"",
+            contents
+        )
+    })?;
+
+    let res = f(index_file).await?;
+    let _ = std::fs::remove_dir_all(&tmp_dir);
```

Comment:
if any error occurs between the time the dir is created and this line; the tmp_dir will still remain (although it would get cleaned up in a future run).
Can use tmpfile; or tmpdir (`tempfile` dependency) - as that would still cleanup the dir if errors occurs in between

Reply:
That is a good catch, i recall you or someone else made the same comment earlier in different fetch function for sources.

I will address all of them together (if you see `git` and `ipfs` sources they also have this happening): I opened #7049, I will add this one too before merging that PR.